File: src/main/scala/leetcode/normal/n930/Solution3.scala
package leetcode.normal.n930

/**
  * Created by CHENXCHEN
  * https://leetcode-cn.com/problems/binary-subarrays-with-sum/
  * 930. 和相同的二元子数组
  *
  * @author <a href="mailto:chenhuachaoxyz@gmail.com">报时</a>
  */
object Solution3 {

  def numSubarraysWithSum(nums: Array[Int], goal: Int): Int = {
    def atMostK(_k: Int): Int = {
      if (_k < 0) return 0
      var res = 0
      var i = 0
      var j = 0
      var k = _k
      while (i < nums.length) {
        k -= nums(i)
        while (k < 0) {
          k += nums(j)
          j += 1
        }
        res += i - j + 1
        i += 1
      }
      res
    }

    atMostK(goal) - atMostK(goal - 1)
  }

}

File: src/main/java/leetcode/normal/j930/Solution3.java
package leetcode.normal.j930;

/**
 * Created by CHENXCHEN
 * https://leetcode-cn.com/problems/binary-subarrays-with-sum/
 * 930. 和相同的二元子数组
 *
 * @author <a href="mailto:chenhuachaoxyz@gmail.com">报时</a>
 */
public class Solution3 {
    public int numSubarraysWithSum(int[] nums, int goal) {
        return atMostK(nums, goal) - atMostK(nums, goal - 1);
    }

    private int atMostK(int[] nums, int k) {
        if (k < 0) return 0;
        int i = 0, j = 0, res = 0;
        for (; i < nums.length; i++) {
            k -= nums[i];
            while (k < 0) {
                k += nums[j];
                j++;
            }
            res += i - j + 1;
        }
        return res;
    }
}
